behaviortree_ros2/test/test_bt_topic_sub_node.cpp

```diff
@@ -0,0 +1,240 @@
+}
+
+TEST_F(TestBtTopicSubNode, TopicFromBlackboard)
+{
+  // GIVEN the input port is remapped to point to a value on the blackboard
+  config_.blackboard->set(kPortIdTopicName, kTopicName);
+
+  // GIVEN the input params do not contain the topic name
+  RosNodeParams params;
+  params.nh = node_;
+
+  // GIVEN we pass in the params and the blackboard when creating the BT node
+  SubNode bt_node("test_node", config_, params);
+
+  // GIVEN we create publisher with default QoS settings after creating the BT node
+  createPublisher(rclcpp::QoS(kHistoryDepth));
+
+  // GIVEN the BT node is RUNNING before the publisher publishes a message
+  ASSERT_THAT(bt_node.executeTick(), testing::Eq(NodeStatus::RUNNING));
+
+  // GIVEN the publisher has published a message
+  publisher_->publish(std_msgs::build<Empty>());
+
+  // WHEN the BT node is ticked
+  // THEN it succeeds
+  EXPECT_THAT(bt_node.executeTick(), testing::Eq(NodeStatus::SUCCESS));
+}
+
+TEST_F(TestBtTopicSubNode, QoSBestEffort)
+{
+  // GIVEN the blackboard does not contain the topic name
+
+  // GIVEN the input params contain the topic name
+  RosNodeParams params;
+  params.nh = node_;
+  params.default_port_value = kTopicName;
+
+  // GIVEN we pass in the params and the blackboard when creating the BT node
+  SubNode bt_node("test_node", config_, params);
+
+  // GIVEN we create publisher with BestEffort reliability QoS settings after creating the BT node
+  createPublisher(rclcpp::QoS(kHistoryDepth).best_effort());
+
+  // GIVEN the BT node is RUNNING before the publisher publishes a message
+  ASSERT_THAT(bt_node.executeTick(), testing::Eq(NodeStatus::RUNNING));
+
+  // GIVEN the publisher has published a message
+  publisher_->publish(std_msgs::build<Empty>());
+
+  // WHEN the BT node is ticked
+  // THEN it succeeds
+  EXPECT_THAT(bt_node.executeTick(), testing::Eq(NodeStatus::SUCCESS));
+}
+
+TEST_F(TestBtTopicSubNode, PublisherCreatedAfterFirstTick)
+{
+  // GIVEN the blackboard does not contain the topic name
+
+  // GIVEN the input params contain the topic name
+  RosNodeParams params;
+  params.nh = node_;
+  params.default_port_value = kTopicName;
+
+  // GIVEN we pass in the params and the blackboard when creating the BT node
+  SubNode bt_node("test_node", config_, params);
+
+  // GIVEN the BT node is RUNNING before the publisher is created
+  ASSERT_THAT(bt_node.executeTick(), testing::Eq(NodeStatus::RUNNING));
+
+  // GIVEN we create publisher with Reliable reliability QoS settings after creating the BT node and after the node starts ticking
+  createPublisher(rclcpp::QoS(kHistoryDepth).reliable());
+
+  // TODO(Joe): Why does the node need to be ticked in between the publisher appearing and sending a message for the message to be received? Seems highly suspicious!
+  ASSERT_THAT(bt_node.executeTick(), testing::Eq(NodeStatus::RUNNING));
+
+  // GIVEN the publisher has published a message
+  publisher_->publish(std_msgs::build<Empty>());
```

Comment:
I would expect to be able to create a publisher and then immediately publish through it without needing to tick the BT node in between. This is an edge case but I think there are situations where it could result in missing messages, such as when manually publishing messages from the command line using `ros2 topic pub`. Does anyone know what could be going on here?

Reply:
the ros executor (which actually processes the receive message) is only spun during the tick action, and when publishing from the command line, as soon as the command is finished the publisher is broken off and no longer available. As ROS2 is peer to peer, if the publisher is gone, it can't send the message through anymore. You can work around this with commandline publishing with the keep alive flag where you can say how long you want the publisher to stay active after the message is published.